// smartsheet/dist/lib/webhooks/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.create = create;
function create(options) {
    const requestor = options.requestor;
    const optionsToSend = Object.assign({}, options.clientOptions);
    const createWebhook = (postOptions, callback) => {
        const urlOptions = { url: buildUrl() };
        return requestor.post(Object.assign(Object.assign(Object.assign({}, optionsToSend), urlOptions), postOptions), callback);
    };
    const deleteWebhook = (deleteOptions, callback) => {
        const urlOptions = { url: buildUrl(deleteOptions.webhookId) };
        return requestor.delete(Object.assign(Object.assign(Object.assign({}, optionsToSend), urlOptions), deleteOptions), callback);
    };
    const updateWebhook = (putOptions, callback) => {
        const urlOptions = { url: buildUrl(putOptions.webhookId) };
        return requestor.put(Object.assign(Object.assign(Object.assign({}, optionsToSend), urlOptions), putOptions), callback);
    };
    const getWebhook = (getOptions, callback) => {
        const urlOptions = { url: buildUrl(getOptions.webhookId) };
        return requestor.get(Object.assign(Object.assign(Object.assign({}, optionsToSend), urlOptions), getOptions), callback);
    };
    const listWebhooks = (getOptions, callback) => {
        const urlOptions = { url: buildUrl() };
        return requestor.get(Object.assign(Object.assign(Object.assign({}, optionsToSend), urlOptions), getOptions), callback);
    };
    const resetSharedSecret = (postOptions, callback) => {
        const urlOptions = { url: buildUrl(postOptions.webhookId) + '/resetsharedsecret' };
        return requestor.post(Object.assign(Object.assign(Object.assign({}, optionsToSend), urlOptions), postOptions), callback);
    };
    const buildUrl = (webhookId) => {
        if (webhookId !== undefined) {
            return options.apiUrls.webhooks + '/' + webhookId;
        }
        return options.apiUrls.webhooks;
    };
    return {
        createWebhook: createWebhook,
        getWebhook: getWebhook,
        listWebhooks: listWebhooks,
        deleteWebhook: deleteWebhook,
        updateWebhook: updateWebhook,
        resetSharedSecret: resetSharedSecret,
    };
}
