// smartsheet/dist/lib/webhooks/types.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.WebhookStatus = exports.WebhookScope = void 0;
// ============================================================================
// Webhook Types
// ============================================================================
/**
 * Webhook scope types
 */
var WebhookScope;
(function (WebhookScope) {
    WebhookScope["SHEET"] = "sheet";
    WebhookScope["PLAN"] = "plan";
})(WebhookScope || (exports.WebhookScope = WebhookScope = {}));
/**
 * Webhook status types
 */
var WebhookStatus;
(function (WebhookStatus) {
    WebhookStatus["DISABLED_ADMINISTRATIVE"] = "DISABLED_ADMINISTRATIVE";
    WebhookStatus["DISABLED_APP_REVOKED"] = "DISABLED_APP_REVOKED";
    WebhookStatus["DISABLED_BY_OWNER"] = "DISABLED_BY_OWNER";
    WebhookStatus["DISABLED_CALLBACK_FAILED"] = "DISABLED_CALLBACK_FAILED";
    WebhookStatus["DISABLED_EXCEEDED_GRID_LIMITS"] = "DISABLED_EXCEEDED_GRID_LIMITS";
    WebhookStatus["DISABLED_SCOPE_INACCESSIBLE"] = "DISABLED_SCOPE_INACCESSIBLE";
    WebhookStatus["DISABLED_VERIFICATION_FAILED"] = "DISABLED_VERIFICATION_FAILED";
    WebhookStatus["ENABLED"] = "ENABLED";
    WebhookStatus["NEW_NOT_VERIFIED"] = "NEW_NOT_VERIFIED";
})(WebhookStatus || (exports.WebhookStatus = WebhookStatus = {}));
