// smartsheet/dist/lib/tokens/index.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.create = create;
const underscore_1 = __importDefault(require("underscore"));
function create(options) {
    const requestor = options.requestor;
    const optionsToSend = {
        url: options.apiUrls.token,
    };
    underscore_1.default.extend(optionsToSend, options.clientOptions);
    delete optionsToSend.accessToken;
    const getAccessToken = (postOptions, callback) => {
        const getTokenQueryStrings = {
            grant_type: 'authorization_code',
        };
        const combinedQueryParams = {
            queryParameters: underscore_1.default.extend(getTokenQueryStrings, postOptions.queryParameters),
        };
        return requestor.post(underscore_1.default.extend({}, optionsToSend, postOptions, combinedQueryParams), callback);
    };
    const refreshAccessToken = (postOptions, callback) => {
        const getTokenQueryStrings = {
            grant_type: 'refresh_token',
        };
        const combinedQueryParams = {
            queryParameters: underscore_1.default.extend(getTokenQueryStrings, postOptions.queryParameters),
        };
        return requestor.post(underscore_1.default.extend({}, optionsToSend, postOptions, combinedQueryParams), callback);
    };
    const revokeAccessToken = (deleteOptions, callback) => {
        const accessTokenOptions = {
            accessToken: options.clientOptions.accessToken,
        };
        return requestor.delete(underscore_1.default.extend({}, optionsToSend, accessTokenOptions, deleteOptions), callback);
    };
    return {
        getAccessToken: getAccessToken,
        refreshAccessToken: refreshAccessToken,
        revokeAccessToken: revokeAccessToken,
    };
}
