// smartsheet/dist/lib/favorites/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.create = create;
const types_1 = require("./types");
function create(options) {
    const requestor = options.requestor;
    const optionsToSend = Object.assign({ url: options.apiUrls.favorites }, options.clientOptions);
    const listFavorites = (getOptions, callback) => requestor.get(Object.assign(Object.assign({}, optionsToSend), getOptions), callback);
    const addItemsToFavorites = (postOptions, callback) => requestor.post(Object.assign(Object.assign({}, optionsToSend), postOptions), callback);
    const handleFavorites = (postOptions, callback) => {
        const body = {
            type: postOptions.type,
            objectId: postOptions.objectId,
        };
        const options = Object.assign(Object.assign({}, postOptions), { body });
        return addItemsToFavorites(options, callback);
    };
    const buildFavoriteAddition = (type) => {
        return (postOptions, callback) => {
            const options = Object.assign(Object.assign({}, JSON.parse(JSON.stringify(postOptions))), { type });
            return handleFavorites(options, callback);
        };
    };
    const addSheetToFavorites = buildFavoriteAddition(types_1.FavoriteType.SHEET);
    const addFolderToFavorites = buildFavoriteAddition(types_1.FavoriteType.FOLDER);
    const addReportToFavorites = buildFavoriteAddition(types_1.FavoriteType.REPORT);
    const addTemplateToFavorites = buildFavoriteAddition(types_1.FavoriteType.TEMPLATE);
    const addWorkspaceToFavorites = buildFavoriteAddition(types_1.FavoriteType.WORKSPACE);
    const addSightToFavorites = buildFavoriteAddition(types_1.FavoriteType.SIGHT);
    const addMultipleToFavorites = (postOptions, callback) => {
        return requestor.post(Object.assign(Object.assign({}, optionsToSend), postOptions), callback);
    };
    const removeFavorite = (deleteOptions, callback) => {
        const favoriteId = deleteOptions.favoriteId;
        const urlOptions = { url: options.apiUrls.favorites + '/' + deleteOptions.favoriteType + '/' + favoriteId };
        return requestor.delete(Object.assign(Object.assign(Object.assign({}, optionsToSend), urlOptions), deleteOptions), callback);
    };
    const removeMultipleFavorites = (deleteOptions, callback) => {
        var _a;
        const urlOptions = {
            url: options.apiUrls.favorites + '/' + deleteOptions.favoriteType,
        };
        // Transform objectIds array to comma-separated string if needed
        const processedOptions = Object.assign({}, deleteOptions);
        if (((_a = processedOptions.queryParameters) === null || _a === void 0 ? void 0 : _a.objectIds) && Array.isArray(processedOptions.queryParameters.objectIds)) {
            processedOptions.queryParameters = Object.assign(Object.assign({}, processedOptions.queryParameters), { objectIds: processedOptions.queryParameters.objectIds.join(',') });
        }
        return requestor.delete(Object.assign(Object.assign(Object.assign({}, optionsToSend), urlOptions), processedOptions), callback);
    };
    const buildFavoriteRemoval = (type) => {
        return (deleteOptions, callback) => {
            const options = JSON.parse(JSON.stringify(deleteOptions));
            options.favoriteType = type;
            return removeFavorite(options, callback);
        };
    };
    const buildMultipleFavoriteRemoval = (type) => {
        return (deleteOptions, callback) => {
            const options = JSON.parse(JSON.stringify(deleteOptions));
            options.favoriteType = type;
            return removeMultipleFavorites(options, callback);
        };
    };
    const removeSheetFromFavorites = buildFavoriteRemoval(types_1.FavoriteType.SHEET);
    const removeFolderFromFavorites = buildFavoriteRemoval(types_1.FavoriteType.FOLDER);
    const removeReportFromFavorites = buildFavoriteRemoval(types_1.FavoriteType.REPORT);
    const removeTemplateFromFavorites = buildFavoriteRemoval(types_1.FavoriteType.TEMPLATE);
    const removeWorkspaceFromFavorites = buildFavoriteRemoval(types_1.FavoriteType.WORKSPACE);
    const removeSightFromFavorites = buildFavoriteRemoval(types_1.FavoriteType.SIGHT);
    return {
        listFavorites,
        addItemsToFavorites,
        addSheetToFavorites,
        addFolderToFavorites,
        addReportToFavorites,
        addTemplateToFavorites,
        addSightToFavorites,
        addWorkspaceToFavorites,
        addMultipleToFavorites,
        removeSheetFromFavorites,
        removeFolderFromFavorites,
        removeReportFromFavorites,
        removeTemplateFromFavorites,
        removeSightFromFavorites,
        removeWorkspaceFromFavorites,
        // Convenience methods to remove multiples.
        removeSheetsFromFavorites: buildMultipleFavoriteRemoval(types_1.FavoriteType.SHEET),
        removeFoldersFromFavorites: buildMultipleFavoriteRemoval(types_1.FavoriteType.FOLDER),
        removeReportsFromFavorites: buildMultipleFavoriteRemoval(types_1.FavoriteType.REPORT),
        removeTemplatesFromFavorites: buildMultipleFavoriteRemoval(types_1.FavoriteType.TEMPLATE),
        removeSightsFromFavorites: buildMultipleFavoriteRemoval(types_1.FavoriteType.SIGHT),
        removeWorkspacesFromFavorites: buildMultipleFavoriteRemoval(types_1.FavoriteType.WORKSPACE),
    };
}
