// smartsheet/dist/lib/users/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.create = create;
const alternateEmails = __importStar(require("./alternateemails"));
function create(options) {
    const requestor = options.requestor;
    const optionsToSend = Object.assign({ url: options.apiUrls.users }, options.clientOptions);
    const listAllUsers = (getOptions, callback) => {
        const requestOptions = Object.assign(Object.assign({}, optionsToSend), getOptions);
        return requestor.get(requestOptions, callback);
    };
    const getUser = (getOptions, callback) => {
        const urlOptions = { url: options.apiUrls.users + '/' + getOptions.userId };
        return requestor.get(Object.assign(Object.assign(Object.assign({}, optionsToSend), urlOptions), getOptions), callback);
    };
    const getCurrentUser = (getOptions, callback) => {
        const urlOptions = { url: options.apiUrls.users + '/me' };
        return requestor.get(Object.assign(Object.assign(Object.assign({}, optionsToSend), urlOptions), getOptions), callback);
    };
    const addUser = (postOptions, callback) => requestor.post(Object.assign(Object.assign({}, optionsToSend), postOptions), callback);
    const addUserAndSendEmail = (postOptions, callback) => addUser(Object.assign(Object.assign(Object.assign({}, optionsToSend), postOptions), { queryParameters: { sendEmail: true } }), callback);
    const updateUser = (putOptions, callback) => {
        const urlOptions = { url: options.apiUrls.users + '/' + putOptions.userId };
        return requestor.put(Object.assign(Object.assign(Object.assign({}, optionsToSend), urlOptions), putOptions), callback);
    };
    const removeUser = (deleteOptions, callback) => {
        const urlOptions = { url: options.apiUrls.users + '/' + deleteOptions.userId };
        return requestor.delete(Object.assign(Object.assign(Object.assign({}, optionsToSend), urlOptions), deleteOptions), callback);
    };
    const deactivateUser = (postOptions, callback) => {
        const urlOptions = { url: buildDeactivateUserUrl(postOptions) };
        return requestor.post(Object.assign(Object.assign(Object.assign({}, optionsToSend), urlOptions), postOptions), callback);
    };
    const reactivateUser = (postOptions, callback) => {
        const urlOptions = { url: buildReactivateUserUrl(postOptions) };
        return requestor.post(Object.assign(Object.assign(Object.assign({}, optionsToSend), urlOptions), postOptions), callback);
    };
    const addProfileImage = (postOptions, callback) => {
        const urlOptions = { url: buildProfileImageUrl(postOptions) };
        return requestor.postFile(Object.assign(Object.assign(Object.assign({}, optionsToSend), urlOptions), postOptions), callback);
    };
    const upgradeUser = (postOptions, callback) => {
        const urlOptions = { url: buildUserUpgradeUrl(postOptions) };
        return requestor.post(Object.assign(Object.assign(Object.assign({}, optionsToSend), urlOptions), postOptions), callback);
    };
    const downgradeUser = (postOptions, callback) => {
        const urlOptions = { url: buildUserDowngradeUrl(postOptions) };
        return requestor.post(Object.assign(Object.assign(Object.assign({}, optionsToSend), urlOptions), postOptions), callback);
    };
    const listUserPlans = (getOptions, callback) => {
        const urlOptions = { url: buildListUserPlansUrl(getOptions) };
        const requestOptions = Object.assign(Object.assign(Object.assign({}, optionsToSend), urlOptions), getOptions);
        return requestor.get(requestOptions, callback);
    };
    const removeUserFromPlan = (deleteOptions, callback) => {
        const urlOptions = { url: buildRemoveUserFromPlanUrl(deleteOptions) };
        const requestOptions = Object.assign(Object.assign(Object.assign({}, optionsToSend), urlOptions), deleteOptions);
        return requestor.delete(requestOptions, callback);
    };
    const buildProfileImageUrl = (urlOptions) => options.apiUrls.users + '/' + urlOptions.userId + '/profileimage';
    const buildDeactivateUserUrl = (urlOptions) => options.apiUrls.users + '/' + urlOptions.userId + '/deactivate';
    const buildReactivateUserUrl = (urlOptions) => options.apiUrls.users + '/' + urlOptions.userId + '/reactivate';
    const buildUserUpgradeUrl = (urlOptions) => options.apiUrls.users + '/' + urlOptions.userId + '/plans/' + urlOptions.planId + '/upgrade';
    const buildUserDowngradeUrl = (urlOptions) => options.apiUrls.users + '/' + urlOptions.userId + '/plans/' + urlOptions.planId + '/downgrade';
    const buildRemoveUserFromPlanUrl = (urlOptions) => options.apiUrls.users + '/' + urlOptions.userId + '/plans/' + urlOptions.planId;
    const buildListUserPlansUrl = (urlOptions) => options.apiUrls.users + '/' + urlOptions.userId + '/plans';
    const userObject = {
        getUser: getUser,
        listAllUsers: listAllUsers,
        getCurrentUser: getCurrentUser,
        addUser: addUser,
        addUserAndSendEmail: addUserAndSendEmail,
        updateUser: updateUser,
        removeUser: removeUser,
        deactivateUser: deactivateUser,
        reactivateUser: reactivateUser,
        addProfileImage: addProfileImage,
        upgradeUser: upgradeUser,
        downgradeUser: downgradeUser,
        listUserPlans: listUserPlans,
        removeUserFromPlan: removeUserFromPlan,
    };
    // Extend with alternate emails functionality
    return Object.assign(Object.assign({}, userObject), alternateEmails.create(options));
}
