// smartsheet/dist/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.smartSheetURIs = exports.createClient = void 0;
const apis_1 = require("./lib/utils/apis");
const index_1 = require("./lib/contacts/index");
const index_2 = require("./lib/events/index");
const index_3 = require("./lib/search/index");
const index_4 = require("./lib/sharing/index");
const index_5 = require("./lib/sights/index");
const httpRequestor_1 = require("./lib/utils/httpRequestor");
const constants = __importStar(require("./lib/utils/constants"));
const index_6 = require("./lib/favorites/index");
const index_7 = require("./lib/folders/index");
const index_8 = require("./lib/groups/index");
const index_9 = require("./lib/home/index");
const index_10 = require("./lib/images/index");
const index_11 = require("./lib/reports/index");
const index_12 = require("./lib/request/index");
const index_13 = require("./lib/server/index");
const index_14 = require("./lib/sheets/index");
const index_15 = require("./lib/templates/index");
const index_16 = require("./lib/tokens/index");
const index_17 = require("./lib/users/index");
const index_18 = require("./lib/webhooks/index");
const index_19 = require("./lib/workspaces/index");
const underscore_1 = __importDefault(require("underscore"));
const winston = __importStar(require("winston"));
// Possible TODO: Namespace parameters for different subcomponents
// E.g. clientOptions.requestor.instance OR
//      clientOptions.requestor.settings
//          w/ sub-paths maxRetryDurationSeconds and calcRetryBackoff
function buildRequestor(clientOptions) {
    if (clientOptions.requestor)
        return clientOptions.requestor;
    const requestorConfig = underscore_1.default.pick(clientOptions, 'maxRetryDurationSeconds', 'calcRetryBackoff');
    if (requestorConfig.maxRetryDurationSeconds)
        requestorConfig.maxRetryDurationMillis = requestorConfig.maxRetryDurationSeconds * 1000;
    requestorConfig.logger = buildLogger(clientOptions);
    return (0, httpRequestor_1.create)(requestorConfig);
}
function buildLogger(clientOptions) {
    if (hasMultipleLogOptions(clientOptions)) {
        throw new Error('Smartsheet client options may specify at most one of ' + "'logger', 'loggerContainer', and 'logLevel'.");
    }
    if (clientOptions.logger)
        return clientOptions.logger;
    if (clientOptions.logLevel)
        return buildLoggerFromLevel(clientOptions.logLevel);
    if (clientOptions.loggerContainer)
        return buildLoggerFromContainer(clientOptions.loggerContainer);
    return null;
}
function hasMultipleLogOptions(clientOptions) {
    return ((clientOptions.logger && clientOptions.loggerContainer) ||
        (clientOptions.logger && clientOptions.logLevel) ||
        (clientOptions.loggerContainer && clientOptions.logLevel));
}
function buildLoggerFromLevel(logLevel) {
    if (winston.level[logLevel] == null) {
        throw new Error('Smartsheet client received configuration with invalid log level ' +
            `'${logLevel}'. Use one of the standard Winston log levels.`);
    }
    return new winston.Logger({
        transports: [
            new winston.transports.Console({
                level: logLevel,
                showLevel: false,
                label: 'Smartsheet',
            }),
        ],
    });
}
function buildLoggerFromContainer(container) {
    if (container.has('smartsheet'))
        return container.get('smartsheet');
    else
        throw new Error('Smartsheet client received a logger container, but could not find a logger named ' + "'smartsheet' inside.");
}
const createClient = function (clientOptions) {
    const requestor = buildRequestor(clientOptions);
    const options = {
        apiUrls: apis_1.apiUrls,
        requestor: requestor,
        clientOptions: {
            accessToken: clientOptions.accessToken || process.env.SMARTSHEET_ACCESS_TOKEN,
            userAgent: clientOptions.userAgent,
            baseUrl: clientOptions.baseUrl,
        },
    };
    return {
        constants: constants,
        contacts: (0, index_1.createContacts)(options),
        events: (0, index_2.createEvents)(options),
        favorites: (0, index_6.create)(options),
        folders: (0, index_7.create)(options),
        groups: (0, index_8.create)(options),
        /**
         * @deprecated
         * The home module is deprecated. The endpoints powering this module
         * are being shut off as part of the sheets folder deprecation.
         * The endpoints will be available until June.
         *
         * See this changelog entry for more information
         * https://developers.smartsheet.com/api/smartsheet/changelog#2025-03-25
         */
        home: (0, index_9.create)(options),
        images: (0, index_10.create)(options),
        reports: (0, index_11.create)(options),
        request: (0, index_12.create)(options),
        search: (0, index_3.createSearch)(options),
        server: (0, index_13.create)(options),
        sharing: (0, index_4.createSharing)(options),
        sheets: (0, index_14.create)(options),
        sights: (0, index_5.createSights)(options),
        templates: (0, index_15.create)(options),
        tokens: (0, index_16.create)(options),
        users: (0, index_17.create)(options),
        webhooks: (0, index_18.create)(options),
        workspaces: (0, index_19.create)(options),
    };
};
exports.createClient = createClient;
exports.smartSheetURIs = {
    defaultBaseURI: 'https://api.smartsheet.com/2.0/',
    govBaseURI: 'https://api.smartsheetgov.com/2.0/',
    euBaseURI: 'https://api.smartsheet.eu/2.0/',
};
__exportStar(require("./lib/events/types"), exports);
__exportStar(require("./lib/sharing/index"), exports);
