// smartsheet/dist/lib/sheets/attachments.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.create = create;
const underscore_1 = __importDefault(require("underscore"));
function create(options) {
    const requestor = options.requestor;
    const optionsToSend = underscore_1.default.extend({}, options.clientOptions);
    const listAttachments = (getOptions, callback) => {
        const urlOptions = { url: buildUrl(getOptions) };
        return requestor.get(underscore_1.default.extend({}, optionsToSend, urlOptions, getOptions), callback);
    };
    const listAttachmentVersions = (getOptions, callback) => {
        const urlOptions = { url: buildUrl(getOptions) + '/versions' };
        return requestor.get(underscore_1.default.extend({}, optionsToSend, urlOptions, getOptions), callback);
    };
    const addUrlAttachment = (postOptions, callback) => {
        const urlOptions = { url: buildUrl(postOptions) };
        return requestor.post(underscore_1.default.extend({}, optionsToSend, urlOptions, postOptions), callback);
    };
    const addFileAttachment = (postOptions, callback) => {
        const urlOptions = { url: buildUrl(postOptions) };
        return requestor.postFile(underscore_1.default.extend({}, optionsToSend, urlOptions, postOptions), callback);
    };
    const attachNewVersion = (postOptions, callback) => {
        const urlOptions = { url: buildUrl(postOptions) + '/versions' };
        return requestor.postFile(underscore_1.default.extend({}, optionsToSend, urlOptions, postOptions), callback);
    };
    const deleteAttachment = (deleteOptions, callback) => {
        const urlOptions = { url: buildUrl(deleteOptions) };
        return requestor.delete(underscore_1.default.extend({}, optionsToSend, urlOptions, deleteOptions), callback);
    };
    const deleteAllAttachmentVersions = (deleteOptions, callback) => {
        const urlOptions = { url: buildUrl(deleteOptions) + '/versions' };
        return requestor.delete(underscore_1.default.extend({}, optionsToSend, urlOptions, deleteOptions), callback);
    };
    const buildUrl = (urlOptions) => {
        let url = options.apiUrls.sheets + '/' + urlOptions.sheetId + '/attachments';
        if (urlOptions.attachmentId !== undefined) {
            url += '/' + urlOptions.attachmentId;
        }
        return url;
    };
    return {
        getAttachment: listAttachments,
        listAttachments: listAttachments,
        listAttachmentVersions: listAttachmentVersions,
        addAttachment: addUrlAttachment,
        addUrlAttachment: addUrlAttachment,
        addFileAttachment: addFileAttachment,
        attachNewVersion: attachNewVersion,
        deleteAttachment: deleteAttachment,
        deleteAllAttachmentVersions: deleteAllAttachmentVersions,
    };
}
