// smartsheet/dist/lib/sheets/columns.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.create = create;
const underscore_1 = __importDefault(require("underscore"));
function create(options) {
    const requestor = options.requestor;
    const optionsToSend = underscore_1.default.extend({}, options.clientOptions);
    const getColumns = (getOptions, callback) => {
        const urlOptions = { url: buildUrl(getOptions) };
        return requestor.get(underscore_1.default.extend({}, optionsToSend, urlOptions, getOptions), callback);
    };
    const addColumn = (postOptions, callback) => {
        const urlOptions = { url: buildUrl(postOptions) };
        return requestor.post(underscore_1.default.extend({}, optionsToSend, urlOptions, postOptions), callback);
    };
    const updateColumn = (putOptions, callback) => {
        const urlOptions = { url: buildUrl(putOptions) };
        return requestor.put(underscore_1.default.extend({}, optionsToSend, urlOptions, putOptions), callback);
    };
    const deleteColumn = (deleteOptions, callback) => {
        const urlOptions = { url: buildUrl(deleteOptions) };
        return requestor.delete(underscore_1.default.extend({}, optionsToSend, urlOptions, deleteOptions), callback);
    };
    const buildUrl = (urlOptions) => {
        let url = options.apiUrls.sheets + '/' + urlOptions.sheetId + '/columns';
        if (urlOptions.columnId !== undefined) {
            url += '/' + urlOptions.columnId;
        }
        return url;
    };
    return {
        getColumns: getColumns,
        getColumn: getColumns,
        addColumn: addColumn,
        deleteColumn: deleteColumn,
        updateColumn: updateColumn,
    };
}
