// smartsheet/dist/lib/sheets/comments.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.create = create;
const underscore_1 = __importDefault(require("underscore"));
function create(options) {
    const requestor = options.requestor;
    const optionsToSend = underscore_1.default.extend({}, options.clientOptions);
    const getComment = (getOptions, callback) => {
        const urlOptions = { url: buildUrl(getOptions) };
        return requestor.get(underscore_1.default.extend({}, optionsToSend, urlOptions, getOptions), callback);
    };
    const deleteComment = (deleteOptions, callback) => {
        const urlOptions = { url: buildUrl(deleteOptions) };
        return requestor.delete(underscore_1.default.extend({}, optionsToSend, urlOptions, deleteOptions), callback);
    };
    const addCommentUrlAttachment = (postOptions, callback) => {
        const urlOptions = { url: buildUrl(postOptions) + '/attachments' };
        return requestor.post(underscore_1.default.extend({}, optionsToSend, urlOptions, postOptions), callback);
    };
    const addCommentFileAttachment = (postOptions, callback) => {
        const urlOptions = { url: buildUrl(postOptions) + '/attachments' };
        return requestor.postFile(underscore_1.default.extend({}, optionsToSend, urlOptions, postOptions), callback);
    };
    const editComment = (putOptions, callback) => {
        const urlOptions = { url: buildUrl(putOptions) };
        return requestor.put(underscore_1.default.extend({}, optionsToSend, urlOptions, putOptions), callback);
    };
    const buildUrl = (urlOptions) => options.apiUrls.sheets + '/' + urlOptions.sheetId + '/comments/' + (urlOptions.commentId || '');
    return {
        getComment: getComment,
        deleteComment: deleteComment,
        addCommentUrlAttachment: addCommentUrlAttachment,
        addCommentAttachment: addCommentUrlAttachment,
        addCommentFileAttachment: addCommentFileAttachment,
        editComment: editComment,
    };
}
