// smartsheet/dist/lib/sheets/create.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.create = create;
const underscore_1 = __importDefault(require("underscore"));
const constants_1 = require("../utils/constants");
function create(options) {
    const requestor = options.requestor;
    const optionsToSend = {
        url: options.apiUrls.sheets,
    };
    underscore_1.default.extend(optionsToSend, options.clientOptions);
    const createSheet = (postOptions, callback) => requestor.post(underscore_1.default.extend({}, optionsToSend, postOptions), callback);
    const createSheetFromExisting = (postOptions, callback) => {
        const options = JSON.parse(JSON.stringify(postOptions));
        if (options.workspaceId) {
            return createSheetInWorkspace(options, callback);
        }
        else if (options.folderId) {
            return createSheetInFolder(options, callback);
        }
        else {
            return createSheet(options, callback);
        }
    };
    const createSheetInFolder = (postOptions, callback) => {
        const urlOptions = { url: buildFolderUrl(postOptions) };
        return requestor.post(underscore_1.default.extend({}, optionsToSend, urlOptions, postOptions), callback);
    };
    const createSheetInWorkspace = (postOptions, callback) => {
        const urlOptions = { url: buildWorkspaceUrl(postOptions) };
        return requestor.post(underscore_1.default.extend({}, optionsToSend, urlOptions, postOptions), callback);
    };
    const copySheet = (postOptions, callback) => {
        const urlOptions = { url: options.apiUrls.sheets + '/' + postOptions.sheetId + '/copy' };
        return requestor.post(underscore_1.default.extend({}, optionsToSend, urlOptions, postOptions), callback);
    };
    const importSheet = (postOptions, callback, contentType, baseUrl) => {
        const urlOptions = {
            url: baseUrl + 'import',
            contentType: contentType,
            contentDisposition: 'attachment',
        };
        return requestor.postFile(underscore_1.default.extend({}, optionsToSend, urlOptions, postOptions));
    };
    const _importXlsxAndReplaceSheet = (postOptions, callback) => {
        const baseUrl = options.apiUrls.sheets + '/' + postOptions.sheetId + '/';
        return importSheet(postOptions, callback, constants_1.acceptHeaders.vndOpenXml, baseUrl);
    };
    const _importCsvAndReplaceSheet = (postOptions, callback) => {
        const baseUrl = options.apiUrls.sheets + '/' + postOptions.sheetId + '/';
        return importSheet(postOptions, callback, constants_1.acceptHeaders.textCsv, baseUrl);
    };
    const importXlsxSheet = (postOptions, callback) => {
        return importSheet(postOptions, callback, constants_1.acceptHeaders.vndOpenXml, options.apiUrls.sheets + '/');
    };
    const importCsvSheet = (postOptions, callback) => {
        return importSheet(postOptions, callback, constants_1.acceptHeaders.textCsv, options.apiUrls.sheets + '/');
    };
    const importXlsxSheetIntoFolder = (postOptions, callback) => {
        return importSheet(postOptions, callback, constants_1.acceptHeaders.vndOpenXml, buildFolderUrl(postOptions) + '/');
    };
    const importCsvSheetIntoFolder = (postOptions, callback) => {
        return importSheet(postOptions, callback, constants_1.acceptHeaders.textCsv, buildFolderUrl(postOptions) + '/');
    };
    const importXlsxSheetIntoWorkspace = (postOptions, callback) => {
        return importSheet(postOptions, callback, constants_1.acceptHeaders.vndOpenXml, buildWorkspaceUrl(postOptions) + '/');
    };
    const importCsvSheetIntoWorkspace = (postOptions, callback) => {
        return importSheet(postOptions, callback, constants_1.acceptHeaders.textCsv, buildWorkspaceUrl(postOptions) + '/');
    };
    const buildFolderUrl = (requestOptions) => {
        return options.apiUrls.folders + '/' + requestOptions.folderId + '/sheets';
    };
    const buildWorkspaceUrl = (requestOptions) => {
        return options.apiUrls.workspaces + '/' + requestOptions.workspaceId + '/sheets';
    };
    return {
        createSheet: createSheet,
        createSheetFromExisting: createSheetFromExisting,
        createSheetInFolder: createSheetInFolder,
        createSheetInWorkspace: createSheetInWorkspace,
        copySheet: copySheet,
        // Not yet released in the API
        // importCsvAndReplaceSheet     : importCsvAndReplaceSheet,
        // importXlsxAndReplaceSheet    : importXlsxAndReplaceSheet,
        importCsvSheet: importCsvSheet,
        importXlsxSheet: importXlsxSheet,
        importCsvSheetIntoFolder: importCsvSheetIntoFolder,
        importXlsxSheetIntoFolder: importXlsxSheetIntoFolder,
        importCsvSheetIntoWorkspace: importCsvSheetIntoWorkspace,
        importXlsxSheetIntoWorkspace: importXlsxSheetIntoWorkspace,
    };
}
