// smartsheet/dist/lib/sheets/discussions.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.create = create;
const underscore_1 = __importDefault(require("underscore"));
function create(options) {
    const requestor = options.requestor;
    const optionsToSend = underscore_1.default.extend({}, options.clientOptions);
    const getDiscussions = (getOptions, callback) => {
        const urlOptions = { url: buildUrl(getOptions) };
        return requestor.get(underscore_1.default.extend({}, optionsToSend, urlOptions, getOptions), callback);
    };
    const listDiscussionAttachments = (getOptions, callback) => {
        const urlOptions = { url: buildUrl(getOptions) + '/attachments' };
        return requestor.get(underscore_1.default.extend({}, optionsToSend, urlOptions, getOptions), callback);
    };
    const createDiscussion = (postOptions, callback) => {
        const urlOptions = { url: buildUrl(postOptions) };
        return requestor.post(underscore_1.default.extend({}, optionsToSend, urlOptions, postOptions), callback);
    };
    const addDiscussionComment = (postOptions, callback) => {
        const urlOptions = { url: buildUrl(postOptions) + '/comments' };
        return requestor.post(underscore_1.default.extend({}, optionsToSend, urlOptions, postOptions), callback);
    };
    const deleteDiscussion = (deleteOptions, callback) => {
        const urlOptions = { url: buildUrl(deleteOptions) };
        return requestor.delete(underscore_1.default.extend({}, optionsToSend, urlOptions, deleteOptions), callback);
    };
    const buildUrl = (urlOptions) => {
        let url = options.apiUrls.sheets + '/' + urlOptions.sheetId + '/discussions';
        if (urlOptions.discussionId !== undefined) {
            url += '/' + urlOptions.discussionId;
        }
        return url;
    };
    return {
        getDiscussions: getDiscussions,
        getDiscussion: getDiscussions,
        listDiscussionAttachments: listDiscussionAttachments,
        createDiscussion: createDiscussion,
        addDiscussionComment: addDiscussionComment,
        deleteDiscussion: deleteDiscussion,
    };
}
