// smartsheet/dist/lib/sheets/get.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.create = create;
const underscore_1 = __importDefault(require("underscore"));
const constants_1 = require("../utils/constants");
function create(options) {
    const requestor = options.requestor;
    const optionsToSend = {
        url: options.apiUrls.sheets,
    };
    underscore_1.default.extend(optionsToSend, options.clientOptions);
    const listSheets = (getOptions, callback) => requestor.get(underscore_1.default.extend({}, optionsToSend, getOptions), callback);
    const getSheetAsCSV = (getOptions, callback) => listSheets(underscore_1.default.extend({}, getOptions, { accept: constants_1.acceptHeaders.textCsv }), callback);
    const getSheetAsPDF = (getOptions, callback) => listSheets(underscore_1.default.extend({}, getOptions, { accept: constants_1.acceptHeaders.applicationPdf, encoding: null }), callback);
    const getSheetAsExcel = (getOptions, callback) => listSheets(underscore_1.default.extend({}, getOptions, { accept: constants_1.acceptHeaders.vndMsExcel, encoding: null }), callback);
    const getSheetVersion = (getOptions, callback) => {
        const urlOptions = { url: options.apiUrls.sheets + '/' + getOptions.sheetId + '/version' };
        return listSheets(underscore_1.default.extend({}, optionsToSend, urlOptions, getOptions), callback);
    };
    const listOrganizationSheets = (getOptions, callback) => {
        const urlOptions = { url: options.apiUrls.users + '/sheets' };
        return requestor.get(underscore_1.default.extend({}, optionsToSend, urlOptions, getOptions), callback);
    };
    return {
        getSheet: listSheets, // will retrieve a single sheet when passed 'id' field
        listSheets: listSheets,
        getSheetAsCSV: getSheetAsCSV,
        getSheetAsExcel: getSheetAsExcel,
        getSheetAsPDF: getSheetAsPDF,
        getSheetVersion: getSheetVersion,
        listOrganizationSheets: listOrganizationSheets,
    };
}
