// smartsheet/dist/lib/sheets/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.create = create;
const share_1 = __importDefault(require("../share/share"));
const underscore_1 = __importDefault(require("underscore"));
const attachments = __importStar(require("./attachments"));
const automationRules = __importStar(require("./automationrules"));
const columns = __importStar(require("./columns"));
const comments = __importStar(require("./comments"));
const createSheets = __importStar(require("./create"));
const crossSheetReferences = __importStar(require("./crosssheetreferences"));
const discussions = __importStar(require("./discussions"));
const getSheets = __importStar(require("./get"));
const summaries = __importStar(require("./sheetsummaries"));
const rows = __importStar(require("./rows"));
const sentUpdateRequests = __importStar(require("./sentupdaterequests"));
const updateRequests = __importStar(require("./updaterequests"));
function create(options) {
    const requestor = options.requestor;
    // Legacy shares module (deprecated)
    const shares = (0, share_1.default)(options.apiUrls.sheets);
    const optionsToSend = {
        urls: options.apiUrls,
    };
    underscore_1.default.extend(optionsToSend, options.clientOptions);
    const updateSheet = (putOptions, callback) => {
        const urlOptions = { url: options.apiUrls.sheets + '/' + putOptions.sheetId };
        return requestor.put(underscore_1.default.extend({}, optionsToSend, urlOptions, putOptions), callback);
    };
    const deleteSheet = (deleteOptions, callback) => {
        const urlOptions = { url: options.apiUrls.sheets + '/' + deleteOptions.sheetId };
        return requestor.delete(underscore_1.default.extend({}, optionsToSend, urlOptions, deleteOptions), callback);
    };
    const sendSheetViaEmail = (postOptions, callback) => {
        const urlOptions = { url: options.apiUrls.sheets + '/' + postOptions.sheetId + '/emails' };
        return requestor.post(underscore_1.default.extend({}, optionsToSend, urlOptions, postOptions), callback);
    };
    const moveSheet = (postOptions, callback) => {
        const urlOptions = { url: options.apiUrls.sheets + '/' + postOptions.sheetId + '/move' };
        return requestor.post(underscore_1.default.extend({}, optionsToSend, urlOptions, postOptions), callback);
    };
    const getPublishStatus = (getOptions, callback) => {
        const urlOptions = { url: options.apiUrls.sheets + '/' + getOptions.sheetId + '/publish' };
        return requestor.get(underscore_1.default.extend({}, optionsToSend, urlOptions, getOptions), callback);
    };
    const setPublishStatus = (putOptions, callback) => {
        const urlOptions = { url: options.apiUrls.sheets + '/' + putOptions.sheetId + '/publish' };
        return requestor.put(underscore_1.default.extend({}, optionsToSend, urlOptions, putOptions), callback);
    };
    const sortRowsInSheet = (postOptions, callback) => {
        const urlOptions = { url: options.apiUrls.sheets + '/' + postOptions.sheetId + '/sort' };
        return requestor.post(underscore_1.default.extend({}, optionsToSend, urlOptions, postOptions), callback);
    };
    const sheetObject = {
        sendSheetViaEmail: sendSheetViaEmail,
        getPublishStatus: getPublishStatus,
        setPublishStatus: setPublishStatus,
        updateSheet: updateSheet,
        deleteSheet: deleteSheet,
        moveSheet: moveSheet,
        sortRowsInSheet: sortRowsInSheet,
    };
    underscore_1.default.extend(sheetObject, attachments.create(options));
    underscore_1.default.extend(sheetObject, automationRules.create(options));
    underscore_1.default.extend(sheetObject, columns.create(options));
    underscore_1.default.extend(sheetObject, comments.create(options));
    underscore_1.default.extend(sheetObject, createSheets.create(options));
    underscore_1.default.extend(sheetObject, crossSheetReferences.create(options));
    underscore_1.default.extend(sheetObject, discussions.create(options));
    underscore_1.default.extend(sheetObject, getSheets.create(options));
    underscore_1.default.extend(sheetObject, summaries.create(options));
    underscore_1.default.extend(sheetObject, rows.create(options));
    underscore_1.default.extend(sheetObject, sentUpdateRequests.create(options));
    underscore_1.default.extend(sheetObject, shares.create(options));
    underscore_1.default.extend(sheetObject, updateRequests.create(options));
    return sheetObject;
}
