// smartsheet/dist/lib/sheets/sheetsummaries.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.create = create;
const underscore_1 = __importDefault(require("underscore"));
function create(options) {
    const requestor = options.requestor;
    const optionsToSend = underscore_1.default.extend({}, options.clientOptions);
    const getSummary = (getOptions, callback) => {
        const urlOptions = { url: buildSummaryUrl(getOptions) };
        return requestor.get(underscore_1.default.extend({}, optionsToSend, urlOptions, getOptions), callback);
    };
    const getSummaryFields = (getOptions, callback) => {
        const urlOptions = { url: buildFieldsUrl(getOptions) };
        return requestor.get(underscore_1.default.extend({}, optionsToSend, urlOptions, getOptions), callback);
    };
    const addSummaryFields = (postOptions, callback) => {
        const urlOptions = { url: buildFieldsUrl(postOptions) };
        return requestor.post(underscore_1.default.extend({}, optionsToSend, urlOptions, postOptions), callback);
    };
    const deleteSummaryFields = (deleteOptions, callback) => {
        const urlOptions = { url: buildFieldsUrl(deleteOptions) };
        return requestor.delete(underscore_1.default.extend({}, optionsToSend, urlOptions, deleteOptions), callback);
    };
    const updateSummaryFields = (putOptions, callback) => {
        const urlOptions = { url: buildFieldsUrl(putOptions) };
        return requestor.put(underscore_1.default.extend({}, optionsToSend, urlOptions, putOptions), callback);
    };
    const addSummaryFieldImage = (postOptions, callback) => {
        const urlOptions = { url: buildFieldImagesUrl(postOptions) };
        return requestor.postFile(underscore_1.default.extend({}, optionsToSend, urlOptions, postOptions), callback);
    };
    const buildSummaryUrl = (urlOptions) => options.apiUrls.sheets + '/' + urlOptions.sheetId + '/summary';
    const buildFieldsUrl = (urlOptions) => buildSummaryUrl(urlOptions) + '/fields';
    const buildFieldImagesUrl = (urlOptions) => buildFieldsUrl(urlOptions) + '/' + urlOptions.fieldId + '/images';
    return {
        getSummary: getSummary,
        getSummaryFields: getSummaryFields,
        addSummaryFields: addSummaryFields,
        deleteSummaryFields: deleteSummaryFields,
        updateSummaryFields: updateSummaryFields,
        addSummaryFieldImage: addSummaryFieldImage,
    };
}
