// smartsheet/dist/lib/sheets/updaterequests.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.create = create;
const underscore_1 = __importDefault(require("underscore"));
function create(options) {
    const requestor = options.requestor;
    const optionsToSend = underscore_1.default.extend({}, options.clientOptions);
    const createUpdateRequest = (postOptions, callback) => {
        const urlOptions = { url: buildUrl(postOptions) };
        return requestor.post(underscore_1.default.extend({}, optionsToSend, urlOptions, postOptions), callback);
    };
    const deleteUpdateRequest = (deleteOptions, callback) => {
        const urlOptions = { url: buildUrl(deleteOptions) };
        return requestor.delete(underscore_1.default.extend({}, optionsToSend, urlOptions, deleteOptions), callback);
    };
    const getUpdateRequest = (getOptions, callback) => {
        const urlOptions = { url: buildUrl(getOptions) };
        return requestor.get(underscore_1.default.extend({}, optionsToSend, urlOptions, getOptions), callback);
    };
    const getAllUpdateRequests = (getOptions, callback) => {
        const urlOptions = { url: buildUrl(getOptions) };
        return requestor.get(underscore_1.default.extend({}, optionsToSend, urlOptions, getOptions), callback);
    };
    const changeUpdateRequest = (putOptions, callback) => {
        const urlOptions = { url: buildUrl(putOptions) };
        return requestor.put(underscore_1.default.extend({}, optionsToSend, urlOptions, putOptions), callback);
    };
    const buildUrl = (urlOptions) => {
        let url = options.apiUrls.sheets + '/' + urlOptions.sheetId + '/updaterequests';
        if (urlOptions.updateRequestId !== undefined) {
            url += '/' + urlOptions.updateRequestId;
        }
        return url;
    };
    return {
        createUpdateRequest: createUpdateRequest,
        deleteUpdateRequest: deleteUpdateRequest,
        getUpdateRequest: getUpdateRequest,
        getAllUpdateRequests: getAllUpdateRequests,
        changeUpdateRequest: changeUpdateRequest,
    };
}
