// smartsheet/dist/lib/sharing/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.createSharing = exports.ShareType = exports.ShareScope = exports.AccessLevel = exports.AssetType = void 0;
/**
 * Asset types supported by the Sharing API
 */
var AssetType;
(function (AssetType) {
    AssetType["SHEET"] = "sheet";
    AssetType["REPORT"] = "report";
    AssetType["SIGHT"] = "sight";
    AssetType["WORKSPACE"] = "workspace";
    AssetType["COLLECTION"] = "collection";
    AssetType["FILE"] = "file";
})(AssetType || (exports.AssetType = AssetType = {}));
/**
 * Access levels for sharing
 */
var AccessLevel;
(function (AccessLevel) {
    AccessLevel["ADMIN"] = "ADMIN";
    AccessLevel["COMMENTER"] = "COMMENTER";
    AccessLevel["EDITOR"] = "EDITOR";
    AccessLevel["EDITOR_SHARE"] = "EDITOR_SHARE";
    AccessLevel["OWNER"] = "OWNER";
    AccessLevel["VIEWER"] = "VIEWER";
})(AccessLevel || (exports.AccessLevel = AccessLevel = {}));
/**
 * Share scope
 */
var ShareScope;
(function (ShareScope) {
    ShareScope["ITEM"] = "ITEM";
    ShareScope["WORKSPACE"] = "WORKSPACE";
})(ShareScope || (exports.ShareScope = ShareScope = {}));
/**
 * Share type
 */
var ShareType;
(function (ShareType) {
    ShareType["GROUP"] = "GROUP";
    ShareType["USER"] = "USER";
})(ShareType || (exports.ShareType = ShareType = {}));
/**
 * Creates and returns the sharing API
 * @param options Create options
 * @returns Sharing API
 */
const createSharing = (options) => {
    const requestor = options.requestor;
    const optionsToSend = Object.assign({}, options.clientOptions);
    // Base URL for shares endpoints
    const baseUrl = '/shares';
    /**
     * List all shares for a specified asset
     * @param options Options for listing shares
     * @param callback Optional callback
     * @returns Promise with shares response
     */
    const listAssetShares = (options, callback) => {
        const { queryParameters } = options;
        // Build the base URL with required parameters
        const urlParams = new URLSearchParams();
        urlParams.append('assetType', queryParameters.assetType.toString());
        urlParams.append('assetId', queryParameters.assetId.toString());
        // Add optional query parameters
        if (queryParameters.maxItems) {
            urlParams.append('maxItems', queryParameters.maxItems.toString());
        }
        if (queryParameters.lastKey) {
            urlParams.append('lastKey', queryParameters.lastKey);
        }
        if (queryParameters.sharingInclude) {
            urlParams.append('sharingInclude', queryParameters.sharingInclude);
        }
        const url = `${baseUrl}?${urlParams.toString()}`;
        return requestor.get(Object.assign(Object.assign(Object.assign({}, optionsToSend), { url }), options), callback);
    };
    /**
     * Get a specific share for a specified asset
     * @param options Options for getting a share
     * @param callback Optional callback
     * @returns Promise with share response
     */
    const getAssetShare = (options, callback) => {
        const { shareId, queryParameters } = options;
        const url = `${baseUrl}/${shareId}?assetType=${queryParameters.assetType}&assetId=${queryParameters.assetId}`;
        return requestor.get(Object.assign(Object.assign(Object.assign({}, optionsToSend), { url }), options), callback);
    };
    /**
     * Share an asset with specified users and/or groups
     * @param options Options for sharing an asset
     * @param callback Optional callback
     * @returns Promise with shares success result
     */
    const shareAsset = (options, callback) => {
        const { body, queryParameters } = options;
        let url = `${baseUrl}?assetType=${queryParameters.assetType}&assetId=${queryParameters.assetId}`;
        if (queryParameters.sendEmail) {
            url += '&sendEmail=true';
        }
        return requestor.post(Object.assign(Object.assign(Object.assign({}, optionsToSend), { url, body }), options), callback);
    };
    /**
     * Update a specific share for a specified asset
     * @param options Options for updating a share
     * @param callback Optional callback
     * @returns Promise with share response
     */
    const updateAssetShare = (options, callback) => {
        const { shareId, queryParameters, body } = options;
        const url = `${baseUrl}/${shareId}?assetType=${queryParameters.assetType}&assetId=${queryParameters.assetId}`;
        return requestor.patch(Object.assign(Object.assign(Object.assign({}, optionsToSend), { url, body }), options), callback);
    };
    /**
     * Delete a specific share for a specified asset
     * @param options Options for deleting a share
     * @param callback Optional callback
     * @returns Promise with success result
     */
    const deleteAssetShare = (options, callback) => {
        const { shareId, queryParameters } = options;
        const url = `${baseUrl}/${shareId}?assetType=${queryParameters.assetType}&assetId=${queryParameters.assetId}`;
        return requestor.delete(Object.assign(Object.assign(Object.assign({}, optionsToSend), { url }), options), callback);
    };
    return {
        listAssetShares,
        getAssetShare,
        shareAsset,
        updateAssetShare,
        deleteAssetShare,
    };
};
exports.createSharing = createSharing;
